README.md

## TheNextQuant

异步事件驱动的量化交易/做市系统。

![](docs/images/framework.png)

![](docs/images/struct.png)


### 框架依赖

- 运行环境
	- python 3.5.3 或以上版本

- 依赖python三方包
	- aiohttp>=3.2.1
	- aioamqp>=0.13.0
	- motor>=2.0.0 (可选)

- RabbitMQ服务器
    - 事件发布、订阅

- MongoDB数据库(可选)
    - 数据存储


### 安装
使用 `pip` 可以简单方便安装:
```text
pip install thenextquant
```

or

```text
pip install -e git+https://github.com/TheNextQuant/thenextquant.git#egg=thenextquant
```

or
```text
pip install . --ignore-installed
```

### Demo使用示例

- 推荐创建如下结构的文件及文件夹:
```text
ProjectName
    |----- docs
    |       |----- README.md
    |----- scripts
    |       |----- run.sh
    |----- config.json
    |----- src
    |       |----- main.py
    |       |----- strategy
    |               |----- strategy1.py
    |               |----- strategy2.py
    |               |----- ...
    |----- .gitignore
    |----- README.md
```

- 快速体验示例
    [Demo](example/demo)


- 运行
```text
python src/main.py config.json
```


### 使用文档

本框架使用的是Python原生异步库(asyncio)实现异步事件驱动，所以在使用之前，需要先了解 [Python Asyncio](https://docs.python.org/3/library/asyncio.html)。

- [服务配置](docs/configure/README.md)
- [行情](docs/market.md)
- [交易](docs/trade.md)
- [资产](docs/asset.md)
- 当前支持交易所
    - [Binance 币币](example/binance)
    - [Binance 合约](example/binance_future)
    - [OKEx 币币](example/okex)
    - [OKEx Margin 杠杆](example/okex_margin)
    - [OKEx Future 交割合约](example/okex_future)
    - [OKEx Swap 永续合约](example/okex_swap)
    - [Deribit 合约](example/deribit)
    - [Bitmex 合约](example/bitmex)
    - [Huobi 币币](example/huobi)
    - [Huobi Future 合约](example/huobi_future)
    - [Coinsuper 币币](example/coinsuper)
    - [Coinsuper Premium 币币](example/coinsuper_pre)
    - [Kraken 币币/杠杆](example/kraken) 
    - [Gate.io 币币](example/gate)
    - [Kucoin 币币](example/kucoin)
    - To be continued ...

- 其它
    - [安装RabbitMQ](docs/others/rabbitmq_deploy.md)
    - [日志打印](docs/others/logger.md)
    - [定时任务](docs/others/tasks.md)

- [框架使用系列课程](https://github.com/TheNextQuant/Documents)


### Change Logs
- [Change Logs](/docs/changelog.md)


### FAQ
- [FAQ](docs/faq.md)


### 有任何问题，欢迎联系

- 微信二维码  
![](docs/images/wx_qrcode.png)


docs/changelog.md
# Change Logs


### v0.2.1

*Date: 2019/10/09*  
*Summary:*
- Add [Binance Future](../example/binance_future) Trade module.


### v0.2.0

*Date: 2019/10/04*  
*Summary:*
- Upgrade OKEx trading module.
- Fix bug: update order remain quantity when status is CANCELED.
- Add volume field in kline data when save to db.
- Add routing_key for EventHeartbeat.
- Change Trade module's locker name to avoid that there was the same locker in one progress.
- Change AMQP login method to PLAIN.


### v0.1.9

*Date: 2019/08/24*  
*Summary:*
- Add [Huobi Future](../example/huobi_future) Trade module.


### v0.1.8

*Date: 2019/08/20*  
*Summary:*
- Upgrade config.PLATFORMS to config.ACCOUNTS.
- validator for string field when field is None will return "".
- logger module: printf stack information in exception.


### v0.1.7

*Date: 2019/08/02*  
*Summary:*
- Add [Kucoin](../example/kucoin) Trade module.
- Add WarningEvent.


### v0.1.6

*Date: 2019/07/29*  
*Summary:*
- Add ConfigEvent to realize run-time-update.
- Add web module, include http server.
- Add validator module.
- Add KeyboardInterrupt caught.
- Enable subscribe multiple events.
- Fix bug: When save kline to mongodb, convert symbol name to collection name.


### v0.1.5

*Date: 2019/07/22*  
*Summary:*
- Add [Coinsuper Premium](../example/coinsuper_pre) Trade module.
- Publish OrderEvent.
- Upgrade Mongodb client.


### v0.1.4

*Date: 2019/07/16*  
*Summary:*
- Add [Gate.io](../example/gate) Trade module.
- Add [Kraken](../example/kraken) Trade module.
- Fix bug for [Coinsuper](../example/coinsuper) : get order infos maybe more than 50 length.


### v0.1.3

*Date: 2019/07/11*  
*Summary:*
- Fix bug: order callback filter by symbol.


### v0.1.2

*Date: 2019/07/04*  
*Summary:*
- fix bug: order & position callback should be use copy.copy() to avoid modified.
- Add [OKEx Swap](../example/okex_swap) Trade module.


### v0.1.1

*Date: 2019/06/28*  
*Summary:*
- Add [Coinsuper](../example/coinsuper) Trade module.


### v0.1.0

*Date: 2019/06/24*  
*Summary:*
- Add [OKEx Margin](../example/okex_margin) Trade module.
- modify EventCenter's Queue name to specific different servers.


### v0.0.9

*Date: 2019/06/14*  
*Summary:*
- Add Asset data callback when create Trade module.
- Add initialize status callback when create Trade module.


### v0.0.8

*Date: 2019/06/03*  
*Summary:*
- Add [Huobi](../example/huobi) module.


### v0.0.7

*Date: 2019/05/31*  
*Summary:*
- Add [Bitmex](https://www.bitmex.com) module.


### v0.0.5

*Date: 2019/05/30*  
*Summary:*
- Add [Binance](../example/binance) module.
- upgrade websocket module.


### v0.0.4

*Date: 2019/05/29*  
*Summary:*
- delete Agent server
- Add [Deribit](../example/deribit) module.
- upgrade market module.


### v0.0.3

*Date: 2019/03/12*  
*Summary:*
- Upgrade Agent Server Protocol to newest version.
- Subscribe Asset.


quant/const.py
# -*- coding:utf-8 -*-

"""
some constants

Author: HuangTao
Date:   2018/07/31
Email:  huangtao@ifclover.com
"""


# Exchange Names
BITFINEX = "bitfinex"
BINANCE = "binance"  # Binance https://www.binance.com
BINANCE_FUTURE = "binance_future"  # https://www.binance-cn.com/cn/futures/BTCUSDT
OKEX = "okex"  # OKEx SPOT https://www.okex.me/spot/trade
OKEX_MARGIN = "okex_margin"  # OKEx MARGIN https://www.okex.me/spot/marginTrade
OKEX_FUTURE = "okex_future"  # OKEx FUTURE https://www.okex.me/future/trade
OKEX_SWAP = "okex_swap"  # OKEx SWAP https://www.okex.me/future/swap
BITMEX = "bitmex"  # BitMEX https://www.bitmex.com/
HUOBI = "huobi"  # Huobi https://www.hbg.com/zh-cn/
HUOBI_FUTURE = "huobi_future"  # Huobi Future https://www.hbdm.com/en-us/contract/exchange/
COINBASE = "coinbase"  # Coinbase https://www.coinbase.com/
MXC = "mxc"
DERIBIT = "deribit"  # Deribit https://www.deribit.com/
KRAKEN = "kraken"  # Kraken https://www.kraken.com
BITSTAMP = "bitstamp"
GEMINI = "gemini"  # Gemini https://gemini.com/
COINSUPER = "coinsuper"  # Coinsuper https://www.coinsuper.com/
COINSUPER_PRE = "coinsuper_pre"  # Coinsuper Premium https://premium.coinsuper.com
GATE = "gate"  # Gate.io https://gateio.news/
KUCOIN = "kucoin"  # Kucoin https://www.kucoin.com/
FCOIN = "fcoin"  # Fcoin https://www.fcoin.com/


# Market Types
MARKET_TYPE_TRADE = "trade"
MARKET_TYPE_ORDERBOOK = "orderbook"
MARKET_TYPE_KLINE = "kline"
MARKET_TYPE_KLINE_5M = "kline_5m"
MARKET_TYPE_KLINE_15M = "kline_15m"
